File: apps/demo-ios/Sources/Views/FuzzCorpus.swift
import Foundation

/// Hardcoded hostile-URL corpus for the §23.2 navigation-fuzzing manual check.
///
/// IMPORTANT — contract note (see README "Fuzz screen vs §7.3 navigation policy"):
/// the Phase-1 core contract (docs/api-contract-ios-v1.md) exposes exactly one
/// public validation entry point, `SEAAuthorizeURLValidator.validate` (§5). The
/// full in-WebView navigation policy (§6 of the contract / spec §7.3) — which is
/// what actually decides allow/block for a URL the WebView is mid-navigation to —
/// is implemented inside SEACore's WKNavigationDelegate and is NOT exposed as a
/// standalone callable API. Since this corpus is evaluated *before* any session
/// starts (there is no WebView to navigate), we run every entry through the one
/// exposed validator. In practice the two policies agree on every entry in this
/// corpus (scheme/host/userinfo/length rules are shared), but this is calling
/// the authorize-URL gate, not the live navigation-delegate code path — flagged
/// here and in the report as a contract ambiguity, not silently papered over.
struct FuzzCorpusEntry: Identifiable {
    let id = UUID()
    let label: String
    let urlString: String
}

enum FuzzCorpus {
    static let entries: [FuzzCorpusEntry] = {
        let overlongQuery = String(repeating: "a=1&", count: 600) // » 2 KB
        return [
            FuzzCorpusEntry(label: "javascript: scheme", urlString: "javascript:alert(1)"),
            FuzzCorpusEntry(
                label: "intent: scheme",
                urlString: "intent://evil/#Intent;scheme=http;package=com.evil;end"
            ),
            FuzzCorpusEntry(label: "file: scheme", urlString: "file:///etc/passwd"),
            FuzzCorpusEntry(label: "plain http (not https)", urlString: "http://auth.bank.local"),
            FuzzCorpusEntry(label: "look-alike host (hyphen prefix)", urlString: "https://evil-auth.bank.local"),
            FuzzCorpusEntry(label: "look-alike host (suffix trick)", urlString: "https://auth.bank.local.evil.io"),
            FuzzCorpusEntry(label: "userinfo trick", urlString: "https://auth.bank.local@evil.io/"),
            FuzzCorpusEntry(label: "data: scheme", urlString: "data:text/html,<script>alert(1)</script>"),
            FuzzCorpusEntry(label: "oversized URL (>2KB)", urlString: "https://auth.bank.local/?\(overlongQuery)"),
        ]
    }()
}

File: apps/demo-android/app/src/main/java/com/bankerise/sea/demo/networking/FuzzCorpus.kt
package com.bankerise.sea.demo.networking

import android.net.Uri
import com.bankerise.sea.core.InvalidUrlReason
import com.bankerise.sea.core.SEAAuthorizeURLValidator
import com.bankerise.sea.core.SEAEnvironment

/**
 * One-tap §23.2 navigation-fuzzing corpus. Each entry is run through
 * [SEAAuthorizeURLValidator.validate] and the result (allow/block + reason)
 * is shown in the Fuzz screen.
 *
 * See the contract-ambiguity note in the iOS demo README: the Fuzz screen
 * exercises the authorize-URL validator, not the live WebViewClient — the
 * two policies should agree but are not the same code path.
 */
object FuzzCorpus {

    data class FuzzEntry(
        val label: String,
        val url: String
    )

    data class FuzzResult(
        val entry: FuzzEntry,
        val allowed: Boolean,
        val reason: InvalidUrlReason?
    )

    private val entries = listOf(
        FuzzEntry("javascript: scheme", "javascript:alert(document.cookie)"),
        FuzzEntry("intent: scheme", "intent://scan/#Intent;scheme=zxing;package=com.google.zxing.client.android;end"),
        FuzzEntry("file: scheme", "file:///data/local/secret"),
        FuzzEntry("content: scheme", "content://media/external/images"),
        FuzzEntry("tel: scheme", "tel:+123456789"),
        FuzzEntry("mailto: scheme", "mailto:someone@example.com"),
        FuzzEntry("plain http", "http://auth-retail.demo.proxym-it.net/realms/test/login"),
        FuzzEntry("uppercase HTTP", "HTTP://auth.bank.com/auth"),
        FuzzEntry("data: scheme", "data:text/html,<script>alert(1)</script>"),
        FuzzEntry("custom scheme (not callback)", "myapp://something"),
        FuzzEntry("callback scheme as non-callback", "BANKERISE-AUTH://callback?code=stolen"),
        FuzzEntry("look-alike host", "https://evil-auth.bank.com/path"),
        FuzzEntry("extra suffix", "https://auth.bank.com.evil.io/path"),
        FuzzEntry("userinfo trick", "https://auth.bank.com@evil.io/path"),
        FuzzEntry("non-standard port", "https://auth.bank.com:8443/path"),
        FuzzEntry("oversized URL", "https://auth.bank.com/${"a".repeat(2050)}"),
        FuzzEntry("valid https", "https://auth-retail.demo.proxym-it.net/test/login"),
        FuzzEntry("localhost (dev)", "https://auth-retail.demo.proxym-it.net/realms/test/login")
    )

    fun run(env: SEAEnvironment, hostAllowlist: List<String>): List<FuzzResult> {
        return entries.map { entry ->
            val uri = Uri.parse(entry.url)
            val result = SEAAuthorizeURLValidator.validate(uri, env, hostAllowlist)
            FuzzResult(
                entry = entry,
                allowed = result.isSuccess,
                reason = result.exceptionOrNull() as? InvalidUrlReason
            )
        }
    }
}
